apps/near_app/near_pub_key.c

```diff
@@ -0,0 +1,421 @@
+      status = false;
+      break;
+    }
+  }
+
+  return status;
+}
+
+static bool get_public_key(const uint8_t *seed,
+                           const uint32_t *path,
+                           uint32_t path_length,
+                           uint8_t *public_key) {
+  HDNode node = {0};
+
+  if (!derive_hdnode_from_path(path, path_length, ED25519_NAME, seed, &node)) {
+    memzero(&node, sizeof(HDNode));
+    return false;
+  }
+
+  if (NULL != public_key) {
+    ecdsa_uncompress_pubkey(
+        get_curve_by_name(ED25519_NAME)->params, node.public_key, public_key);
+  }
+
+  memzero(&node, sizeof(HDNode));
+  return true;
+}
+
+static void fill_public_keys(const near_get_public_key_derivation_path_t *paths,
+                             const uint8_t *seed,
+                             uint8_t public_key_list[][NEAR_PUB_KEY_SIZE],
+                             pb_size_t count) {
+  for (pb_size_t index = 0; index < count; index++) {
+    const near_get_public_key_derivation_path_t *path = &paths[index];
+    if (!get_public_key(
+            seed, path->path, path->path_count, public_key_list[index])) {
+      // send unknown error; unknown failure reason
+      near_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static bool send_public_keys(near_query_t *query,
+                             const uint8_t pubkey_list[][NEAR_PUB_KEY_SIZE],
+                             const pb_size_t count,
+                             const pb_size_t which_request) {
+  near_result_t response = init_near_result(NEAR_RESULT_GET_PUBLIC_KEY_TAG);
+  near_get_public_key_result_response_t *result =
+      &response.get_public_key.result;
+  size_t batch_limit =
+      sizeof(response.get_public_keys.result.public_key) / NEAR_PUB_KEY_SIZE;
+  size_t remaining = count;
+
+  response.get_public_keys.which_response =
+      NEAR_GET_PUBLIC_KEY_RESPONSE_RESULT_TAG;
+  while (true) {
+    // send response as batched list of public keys
+    size_t batch_size = CY_MIN(batch_limit, remaining);
+    result->public_key_count = batch_size;
+
+    memcpy(result->public_key,
+           &pubkey_list[count - remaining],
+           batch_size * NEAR_PUB_KEY_SIZE);
+
+    near_send_result(&response);
+    remaining -= batch_size;
+    if (0 == remaining) {
+      break;
+    }
+
+    if (!near_get_query(query, which_request) ||
+        !check_which_request(query,
+                             NEAR_GET_PUBLIC_KEY_REQUEST_FETCH_NEXT_TAG)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static bool get_user_consent(const pb_size_t which_request,
+                             const char *wallet_name) {
+  char msg[100] = "";
+
+  if (NEAR_QUERY_GET_PUBLIC_KEYS_TAG == which_request) {
+    snprintf(
+        msg, sizeof(msg), UI_TEXT_ADD_ACCOUNT_PROMPT, NEAR_SYMBOL, wallet_name);
+  } else {
+    snprintf(
+        msg, sizeof(msg), UI_TEXT_RECEIVE_PROMPT, NEAR_SYMBOL, wallet_name);
```

Comment:
Verify if `UI_TEXT_RECEIVE_TOKEN_PROMPT` should be used instead of `UI_TEXT_RECEIVE_PROMPT`.

`UI_TEXT_RECEIVE_TOKEN_PROMPT` is used in the previous implementation (check in controller_main.c #734)

Reply:
Got it, in original implementation `NEAR` is treated like a token and 'Near' as the chain name.

I will fix it.